ledger/participant-integration-api/src/main/scala/platform/store/FlywayMigrationsBase.scala

```diff
@@ -0,0 +1,294 @@
+    loggingContext: LoggingContext,
+) {
+  private val logger = ContextualizedLogger.get(this.getClass)
+  implicit private val ec: ExecutionContext = resourceContext.executionContext
+
+  // A "baseline migration" represents all migrations up to and including the "baseline version"
+  // Baseline migrations are a first class feature in Flyway, however they are not available in the free version.
+  // To move the baseline version, use the following procedure:
+  //   1. Apply all migrations up to and including the new baseline version to a fresh database
+  //   2. Export the resulting schema as the new baseline migration, and assign it a version number equal to the
+  //      baseline version
+  //   3. Double check that the migrations from step 1 and the migration from step 2 produce exactly the same
+  //      database
+  //   4. Delete all migrations used in step 1
+  //   5. Update the value below
+  protected def baselineVersion: String
+
+  protected def getDataSource: Future[DataSource]
+
+  protected def locations: List[String]
+
+  private def runF[T](t: FluentConfiguration => Future[T]): Future[T] =
+    getDataSource.flatMap(dataSource => t(configurationBase(dataSource)))
+
+  private def run[T](t: FluentConfiguration => T): Future[T] =
+    runF(fc => Future(t(fc)))
+
+  private def configurationBase(dataSource: DataSource): FluentConfiguration =
+    Flyway
+      .configure()
+      .locations(locations: _*)
+      .dataSource(dataSource)
+
+  private def getStatus(flyway: Flyway): FlywayStatus = {
+    val info = flyway.info()
+    val current = info.current()
+    val pending = info.pending()
+    val applied = info.applied()
+
+    def isBeforeBaseline(m: MigrationInfo) = !m.getVersion.isAtLeast(baselineVersion)
+
+    // The Flyway history is immutable.
+    // To delete a migration, a new row is inserted into the history table that marks the migration as deleted.
+    // We are looking for rows that have not been marked as deleted by other rows (not DELETED),
+    // and that are not themselves deletion markers (not DELETE).
+    val notDeletedMigrationsBeforeBaseline = applied.toList
+      .filter(m =>
+        isBeforeBaseline(
+          m
+        ) && m.getType != MigrationType.DELETE && m.getState != MigrationState.DELETED
+      )
+      .map(_.getVersion.getVersion)
+
+    if (current == null) {
+      FlywayStatus.DatabaseEmpty(info.pending().length)
+    } else if (isBeforeBaseline(current)) {
+      FlywayStatus.SchemaBeforeBaseline(current.getVersion.getVersion, baselineVersion)
+    } else if (notDeletedMigrationsBeforeBaseline.nonEmpty) {
+      FlywayStatus.NeedsBaselineReset(notDeletedMigrationsBeforeBaseline)
+    } else if (pending.nonEmpty) {
+      FlywayStatus.Outdated(current.getVersion.getVersion, applied.length, pending.length)
+    } else {
+      FlywayStatus.NoMigrationsPending(current.getVersion.getVersion, applied.length)
+    }
+  }
+
+  private def resetBaseline(flyway: Flyway): Unit = {
```

Comment:
repearing is not really a migration operation: did you consider doing this anyway? like a pre-step? it would make it possible to have less combinations, and some more decoupling.